### DISCUSSION TITLE
Questions about Tasklist and calendar

### DISCUSSION THREAD

**Discussion by ghost** [awesomeWM/awesome#3900] [Q&A]
Hello, I have 2 questions:

1. I am using this calendar:
![Screenshot_20240222_22-48-27](https://github.com/awesomeWM/awesome/assets/136118735/54c31cf7-d4dd-4250-b6d1-5af62c111b97)

```
mytextcalendar = wibox.widget.textclock("%d %b %Y", 2, "Europe/London")

local function rounded_shape(size, partial)
    if partial then
        return function(cr, width, height)
                   gears.shape.partially_rounded_rect(cr, width, height,
                        false, true, false, true, 5)
               end
    else
        return function(cr, width, height)
                   gears.shape.rounded_rect(cr, width, height, size)
               end
    end
end

local month_calendar = awful.widget.calendar_popup.year({
                 font = "monaco 11", 
                 opacity=0.8,
                 spacing=0,
                 margin=0,
                 style_month   = { padding      = 0,
                                   bg_color     = '#222222',
                                   border_width = 3,
                                   border_color = '#222222',
                                   shape        = rounded_shape(10)},
                 style_normal  = { shape    = rounded_shape(5), },
                 style_focus   = { fg_color = '#222222',
                                   bg_color = '#ff9800',
                                   shape    = rounded_shape(5, true)},
                 style_header  = { fg_color = '#ff9800',
                                   border_width = 0,
                                   shape    = rounded_shape(10)},
                                   
                 style_yearheader = {border_width = 0,
                                     bg_color = '#222222',
                                     fg_color = '#ff9800'},
                                     
                 style_weekday = { fg_color = '#de5e1e',
                                   border_width = 0,
                                   shape    = rounded_shape(5)}
                 })
month_calendar:attach( mytextcalendar, "tr", { on_hover = false, })
```

How can I colour Sundays and Saturdays in a slightly different colour please? I.e. like this one:

https://awesomewm.org/doc/api/classes/wibox.widget.calendar.html#wibox.widget.calendar.fn_embed

Another thing is: Is there a way to disable the mouse scroll key to change the year (it moves multiple years at a time), for example use arrow keys instead?

***

2. The tasklist widget layout can be set to either `wibox.layout.flex.horizontal()` or `wibox.layout.fixed.horizontal()`. Is it possible to set it to fixed, then only when it is full, change this setting to flexible? Thanks.

**Comment by actionless** [comment]
> How can I colour Sundays and Saturdays in a slightly different colour please? I.e. like this one:
> https://awesomewm.org/doc/api/classes/wibox.widget.calendar.html#wibox.widget.calendar.fn_embed

it's literally specified by that link which you provided yourself:
![2024-03-16--1710563501_1129x1766_scrot](https://github.com/awesomeWM/awesome/assets/1655669/47a38496-4bdb-4b28-80fd-6b58b1f3b8d7)

**Reply by ghost** [reply]
I did this:

```
mytextcalendar = wibox.widget.textclock("%d %b %Y", 2, "Europe/London")

local function rounded_shape(size, partial)
    if partial then
        return function(cr, width, height)
                   gears.shape.partially_rounded_rect(cr, width, height,
                        false, true, false, true, 5)
               end
    else
        return function(cr, width, height)
                   gears.shape.rounded_rect(cr, width, height, size)
               end
    end
end

local function decorate_cell(widget, flag, date)
    if flag=='monthheader' and not styles.monthheader then
        flag = 'header'
    end
    local props = styles[flag] or {}
    if props.markup and widget.get_text and widget.set_markup then
        widget:set_markup(props.markup(widget:get_text()))
    end
    -- Change bg color for weekends
    local d = {year=date.year, month=(date.month or 1), day=(date.day or 1)}
    local weekday = tonumber(os.date('%w', os.time(d)))
    local default_bg = (weekday==0 or weekday==6) and '#232323' or '#383838'
    local ret = wibox.widget {
        {
            widget,
            margins = (props.padding or 2) + (props.border_width or 0),
            widget  = wibox.container.margin
        },
        shape              = props.shape,
        shape_border_color = props.border_color or '#b9214f',
        shape_border_width = props.border_width or 0,
        fg                 = '#999999',
        bg                 = '#000000',
        widget             = wibox.container.background
    }
    return ret
end

local month_calendar = awful.widget.calendar_popup.year({
                 font = "monaco 11", 
                 opacity=0.8,
                 spacing=0,
                 margin=0,
                 fn_embed = decorate_cell, -- <=========================================
                 style_month   = { padding      = 0,
                                   bg_color     = '#222222',
                                   border_width = 3,
                                   border_color = '#222222',
                                   shape        = rounded_shape(10)},
                 style_normal  = { shape    = rounded_shape(5), },
                 style_focus   = { fg_color = '#222222',
                                   bg_color = '#ff9800',
                                   shape    = rounded_shape(5, true)},
                 style_header  = { fg_color = '#ff9800',
                                   border_width = 0,
                                   shape    = rounded_shape(10)},
                                   
                 style_yearheader = {border_width = 0,
                                     bg_color = '#222222',
                                     fg_color = '#ff9800'},
                                     
                 style_weekday = { fg_color = '#de5e1e',
                                   border_width = 0,
                                   shape    = rounded_shape(5)}
                 })
month_calendar:attach( mytextcalendar, "tr", { on_hover = false, })
```

But it doesn't work.

**Reply by actionless** [reply]
does it works if you just copy-paste it from the example, ie:

![2024-06-28--1719573254_388x108_scrot](https://github.com/awesomeWM/awesome/assets/1655669/161d9a0d-5e38-4043-a2ff-13b077a5fa45)

?

**Reply by actionless** [reply]
because you're using calendar_popup, not calendar, it's a different widget, with a different documentation: https://awesomewm.org/doc/api/classes/awful.widget.calendar_popup.html

**Reply by ghost** [reply]
Thank you. I see, this makes sense.

> ![2024-06-28--1719573254_388x108_scrot](https://private-user-images.githubusercontent.com/1655669/344141940-161d9a0d-5e38-4043-a2ff-13b077a5fa45.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MTk1NzU0MjcsIm5iZiI6MTcxOTU3NTEyNywicGF0aCI6Ii8xNjU1NjY5LzM0NDE0MTk0MC0xNjFkOWEwZC01ZTM4LTQwNDMtYTJmZi0xM2IwNzdhNWZhNDUucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI0MDYyOCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNDA2MjhUMTE0NTI3WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9NGY3MjUxN2ZhMzNjZDcyNDYwM2ZhNmNkZjY1ODk1Y2RiZWJjMTZkNWQyMDFkYzM3ZjVmMDZhNDNhNmM5NTE4MyZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QmYWN0b3JfaWQ9MCZrZXlfaWQ9MCZyZXBvX2lkPTAifQ.4L9IbNbxk4lKtz_BX_mBJsr7LFm90J4YGKvc-xqJDQw)

How can you use the variable `cal`? What do I do with it? Linking it to popup like this does not work:
`cal:attach( mytextcalendar, "tr", { on_hover = true, })`

**Reply by actionless** [reply]
>  this does not work:

of course it won't - such method not documented for it to work

see the source of calendar_popup - you can copy it to your local config, and modify to use your customized calendar widget

also feel free to make a PR to expose fn_embed argument into calendar_popup widget
